### PR TITLE
impl(oauth2): get universe_domain field from json credentials

### LINKED ISSUE #13426: Parse `universe_domain` field from JSON

### PR DESCRIPTION
**Pull request by scotthart**
fixes #13426

This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13321)

### CODE REVIEW

File: google/cloud/internal/oauth2_universe_domain_test.cc [hunk 1]
```diff
@@ -0,0 +1,101 @@
+// Copyright 2023 Google LLC
```
Comment (coryan) [inline google/cloud/internal/oauth2_universe_domain_test.cc:1]: Here and in any new file s/2023/2024/

File: google/cloud/internal/oauth2_universe_domain.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_OAUTH2_UNIVERSE_DOMAIN_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_OAUTH2_UNIVERSE_DOMAIN_H
+
+#include "google/cloud/options.h"
+#include "google/cloud/status.h"
+#include "google/cloud/status_or.h"
+#include "google/cloud/version.h"
+#include <nlohmann/json_fwd.hpp>
```
Comment (coryan) [inline google/cloud/internal/oauth2_universe_domain.h:22]: TIL

File: google/cloud/internal/oauth2_minimal_iam_credentials_rest.cc [hunk 3]
```diff
@@ -21,6 +21,7 @@
 #include "google/cloud/internal/json_parsing.h"
 #include "google/cloud/internal/make_status.h"
 #include "google/cloud/internal/oauth2_credentials.h"
+#include "google/cloud/internal/oauth2_universe_domain.h"
```
Comment (coryan) [inline google/cloud/internal/oauth2_minimal_iam_credentials_rest.cc:24]: I this used?

**Review by coryan — COMMENTED** [review]
All looks good except the copyright dates.  I was expecting changes for the compute engine credentials, I assume that will happen separately?

It may have been nicer to split this a bit more

**Review by scotthart — COMMENTED** [review]
Yes. Metadata server interaction will be in a future PR.

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13321#-:-NnoNVzVEnmEvocOl0lG:b-13krbo)* status: 0 of 22 files reviewed, 3 unresolved discussions (waiting on @coryan)

___
*[`google/cloud/internal/oauth2_minimal_iam_credentials_rest.cc` line 24 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13321#gh-1447469636:-NnoNeo-3kk9hbEOHO7a:bhsmxy7) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/ed56d8c7a570c806834a32af0265284cb0e781af/google/cloud/internal/oauth2_minimal_iam_credentials_rest.cc#L24)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I this used?
</blockquote></details>

Not any longer.
___
*[`google/cloud/internal/oauth2_universe_domain.h` line 22 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13321#gh-1447465834:-NnoNi-o30DqArpu5BNA:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/ed56d8c7a570c806834a32af0265284cb0e781af/google/cloud/internal/oauth2_universe_domain.h#L22)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

TIL
</blockquote></details>

Done.
___
*[`google/cloud/internal/oauth2_universe_domain_test.cc` line 1 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13321#gh-1447464889:-NnoNid7ABwBMhj3Lmcu:bt2vr1b) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/ed56d8c7a570c806834a32af0265284cb0e781af/google/cloud/internal/oauth2_universe_domain_test.cc#L1)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Here and in any new file s/2023/2024/
</blockquote></details>

Fixed.

**Review by coryan — APPROVED** [review]
> Reviewed 22 of 22 files at r1, 3 of 3 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13321)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @scotthart)

**Pull request merged**